pkg/controllers/resources/persistentvolumes/syncer.go

```diff
@@ -129,10 +129,17 @@ func (s *persistentVolumeSyncer) Sync(ctx *synccontext.SyncContext, pObj client.
 
 	// check if virtual persistent volume is deleted
 	if vPersistentVolume.GetDeletionTimestamp() != nil && len(vPersistentVolume.GetFinalizers()) > 0 {
-		//delete the finalizer, so that the object can be deleted
-		vPersistentVolume.SetFinalizers(nil)
-		ctx.Log.Infof("remove virtual persistent volume %s finalizers, because virtual persistent volume is terminating", vPersistentVolume.GetName())
-		return ctrl.Result{}, ctx.VirtualClient.Update(ctx.Context, vPersistentVolume)
+		if vPersistentVolume.Spec.ClaimRef != nil && vPersistentVolume.Spec.PersistentVolumeReclaimPolicy == corev1.PersistentVolumeReclaimDelete {
+			// executes when dynamically provisioned PV with a delete reclaim policy gets deleted
+			// it also releases the cloud provider volume.
+			vPersistentVolume.SetFinalizers(nil)
```

Comment:
I am not getting why the finalizer should be removed by the syncer. Could you explain that for me, please?

My initial instinct tells me to not touch the finalizers, and just wait until CSI driver or PV controller deletes the finalizer. If the reclaim policy is "Delete", that to me means that the driver needs to confirm the deletion from the filesystem(or cloud API) before deleting the PV in k8s - finalizers are key part of that process. If we remove the finalizer and PV is stuck in the deleting state, the vcluster user won't know about it, and PV in the host cluster will still exist.

Reply:
Hi @matskiv. As per the steps explained in the Issue #600, the SC (Reclaim Policy = Delete), PVC and Pod are created on the virtual cluster, the PV gets dynamically provisioned. Now when the vPod and vPVC are deleted, the syncer triggers the deletion of pPVC; and host cluster initiates the deletion of the pPV and the cloud provider volume. But only the vPV is stuck in the 'Terminating' state. Hence, we are clearing the finalizers so that the same gets garbage collected.

Reply:
But if the pPV is deleted, then the SyncDown() is executed, but not the Sync(). And in SyncDown() we have finalizers removal already implemented.

Reply:
Before this fix. when the vPVC was deleted, many events occured. Out of which the following can be noted:

1. the Sync() gets called first where the vPV was deleted, vPV & pPV deletion timestamp were checked and pPV was deleted using PhysicalClient.Delete(). Sync() gets called repeatedly and the reconcile request is re-queued until pPV got deleted from the host cluster.
2. the SyncDown() was called when the pPV was deleted; the vPV deletion timestamp was checked and finalizers were removed.

Even though pPV and vPV got deleted, the cloud provider volume deletion was intermittent. So, I think the call to PhysicalClient.Delete() may not be required, since the host cluster takes care of the deletion of the pPV in the case of a dynamically provisioned PV with a reclaim policy of 'Delete'. 

With this fix, when the finalizers are cleared in the Sync(), a call to SyncDown() is not required because the pPV has already been deleted by the host cluster.

Reply:
The previous behavior seems right to me. Why should the vPV be deleted before the pPV is actually gone? If the pPV deletion fails(or gets stuck), the user will never find out.

I still don't understand how deleting finalizers on the vPV helps to ensure that the cloud provider actually deletes the volume. :/

Reply:
@matskiv I have pushed another commit wherein the clearing of the finalizers have been avoided in Sync() and instead we wait for the physical PV to be deleted naturally and let the SyncDown() clear the finalizers on the virtual PV. Kindly let me know of the code changes are OK.